pkg/supportbundle/supportbundle.go

```diff
@@ -0,0 +1,310 @@
+package supportbundle
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/pprof"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/pkg/build"
+	"github.com/grafana/agent/pkg/config"
+	"github.com/grafana/agent/pkg/traces"
+	zaplogfmt "github.com/jsternberg/zap-logfmt"
+	"github.com/mackerelio/go-osstat/uptime"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"gopkg.in/yaml.v3"
+)
+
+// Bundle collects all the data that is exposed as a support bundle.
+type Bundle struct {
+	meta                  []byte
+	config                []byte
+	agentMetrics          []byte
+	agentMetricsInstances []byte
+	agentMetricsTargets   []byte
+	agentLogsInstances    []byte
+	heapBuf               *bytes.Buffer
+	goroutineBuf          *bytes.Buffer
+	blockBuf              *bytes.Buffer
+	mutexBuf              *bytes.Buffer
+	cpuBuf                *bytes.Buffer
+}
+
+// Metadata contains general runtime information about the current Agent.
+type Metadata struct {
+	BuildVersion    string   `yaml:"build_version"`
+	OS              string   `yaml:"os"`
+	Architecture    string   `yaml:"architecture"`
+	Uptime          float64  `yaml:"uptime"`
+	EnabledFeatures []string `yaml:"enabled_features"`
+}
+
+// Export gathers the information required for the support bundle.
+func Export(enabledFeatures []string, cfg config.Config, srvAddress string, duration float64) (*Bundle, error) {
```

Comment:
Would it make sense to split this function a bit?

Reply:
Sorry for not responding for so long, I refactored some things about the PR, but now getting to the supportbundle package itself. What do you have in mind?